gdal/ogr/ogrsf_frmts/flatgeobuf/geometryreader.cpp

```diff
@@ -232,212 +226,165 @@ OGRErr GeometryReader::readSimpleCurve(OGRSimpleCurve *sc)
     for (uoffset_t i = 0; i < parts->size(); i++) {
         GeometryReader reader { parts->Get(i), m_hasZ, m_hasM };
-        auto geom = reader.read();
-        auto curve = dynamic_cast<OGRCurve*>(geom);
-        if( curve == nullptr )
-        {
-            delete geom;
-            delete cp;
+        auto c = std::unique_ptr<OGRCurve>(reader.read()->toCurve());
+        if (c == nullptr)
             return nullptr;
-        }
-        cp->addRingDirectly(curve);
+        cp->addRingDirectly(c.release());
     }
-    return cp;
+    return cp.release();
 }
 
 OGRMultiCurve *GeometryReader::readMultiCurve()
 {
     auto parts = m_geometry->parts();
     if (parts == nullptr)
         return CPLErrorInvalidPointer("parts data");
-    auto mc = new OGRMultiCurve();
+    auto mc = std::unique_ptr<OGRMultiCurve>(new OGRMultiCurve());
     for (uoffset_t i = 0; i < parts->size(); i++) {
         GeometryReader reader { parts->Get(i), m_hasZ, m_hasM };
-        auto geom = reader.read();
-        auto curve = dynamic_cast<OGRCurve*>(geom);
-        if( curve == nullptr )
-        {
-            delete geom;
-            delete mc;
+        auto c = std::unique_ptr<OGRCurve>(reader.read()->toCurve());
+        if (c == nullptr)
             return nullptr;
-        }
-        mc->addGeometryDirectly(curve);
+        mc->addGeometryDirectly(c.release());
     }
-    return mc;
+    return mc.release();
 }
 
 OGRMultiSurface *GeometryReader::readMultiSurface()
 {
     auto parts = m_geometry->parts();
     if (parts == nullptr)
         return CPLErrorInvalidPointer("parts data");
-    auto ms = new OGRMultiSurface();
+    auto ms = std::unique_ptr<OGRMultiSurface>(new OGRMultiSurface());
     for (uoffset_t i = 0; i < parts->size(); i++) {
         GeometryReader reader { parts->Get(i), m_hasZ, m_hasM };
-        auto geom = reader.read();
-        auto surf = dynamic_cast<OGRSurface*>(geom);
-        if( surf == nullptr )
-        {
-            delete surf;
-            delete ms;
+        auto s = std::unique_ptr<OGRSurface>(reader.read()->toSurface());
+        if (s == nullptr)
             return nullptr;
-        }
-        ms->addGeometryDirectly(surf);
+        ms->addGeometryDirectly(s.release());
     }
-    return ms;
+    return ms.release();
 }
 
 OGRPolyhedralSurface *GeometryReader::readPolyhedralSurface()
 {
     auto parts = m_geometry->parts();
     if (parts == nullptr)
         return CPLErrorInvalidPointer("parts data");
-    auto ps = new OGRPolyhedralSurface();
+    auto ps = std::unique_ptr<OGRPolyhedralSurface>(new OGRPolyhedralSurface());
     for (uoffset_t i = 0; i < parts->size(); i++) {
         GeometryReader reader { parts->Get(i), m_hasZ, m_hasM };
-        auto geom = reader.read();
-        auto surf = dynamic_cast<OGRSurface*>(geom);
-        if( surf == nullptr )
-        {
-            delete surf;
-            delete ps;
+        auto s = std::unique_ptr<OGRSurface>(reader.read()->toSurface());
```

Comment:
this change is not equivent to previous code. the dynamic_cast<OGRSurface*>(geom) is (I think) needed to make sure the sub-object was really a OGRSurface. toSurface() resolves to a static_cast<> in non-debug builds, and so will result in a incorrect pointer if the geometry was not a Surface